pkg/client/interface.go

```diff
@@ -94,33 +99,34 @@ type TxContext interface {
 	) (*comettypes.ResultTx, error)
 }
 
-// BlocksObservable is an observable which is notified with an either
-// value which contains either an error or the event message bytes.
-//
-// TODO_HACK: The purpose of this type is to work around gomock's lack of
-// support for generic types. For the same reason, this type cannot be an
-// alias (i.e. EventsBytesObservable = observable.Observable[either.Either[[]byte]]).
-type BlocksObservable observable.ReplayObservable[Block]
-
-// BlockClient is an interface which provides notifications about newly committed
-// blocks as well as direct access to the latest block via some blockchain API.
-type BlockClient interface {
-	// CommittedBlocksSequence returns an observable which emits newly committed blocks.
-	CommittedBlocksSequence(context.Context) BlocksObservable
-	// LatestBlock returns the latest block that has been committed.
-	LatestBlock(context.Context) Block
-	// Close unsubscribes all observers of the committed blocks sequence observable
-	// and closes the events query client.
-	Close()
-}
-
 // Block is an interface which abstracts the details of a block to its minimal
 // necessary components.
 type Block interface {
 	Height() int64
 	Hash() []byte
 }
 
+// DelegateeChange is an interface which wraps the EventDelegateeChange event
+// emitted by the application module.
+type DelegateeChange interface {
+	AppAddress() string
+}
+
+// EventsObservable is a replay observable for events of some type T.
+type EventsObservable[T any] observable.ReplayObservable[T]
+
+// MappedClient is an interface which provides notifications about newly received
+// events as well as direct access to the latest event via some blockchain API.
+type MappedClient[T any] interface {
+	// EventsSequence returns an observable which emits new events.
+	EventsSequence(context.Context) EventsObservable[T]
+	// LatestEvent returns the latest event that has been received.
+	LatestEvent(context.Context) T
+	// Close unsubscribes all observers of the events sequence observable
+	// and closes the events query client.
+	Close()
+}
```

Comment:
These is the generic MappedClient interface

Reply:
@h5law After thinking about it for a bit, I think this should be called `EventsReplayClient`.

Reply:
I still plan on looking more closely at the package structure but haven't had a chance to do so yet.

Reply:
I actually called this the ReplayClient first then the TransformingClient 🤖 then MappedClient lmao

Reply:
I also think that the block and delegate clients should lightly wrap (as opposed to bring or embedding) this generic type so that they have control over their own method names and, more importantly, become open for extension (S**O**LID).